### ISSUE TITLE
Support for international characters and UTF8

### ISSUE THREAD

**Issue by mobluse** [boxgaming/qbjs#114]
I cannot get Swedish characters, åäöÅÄÖ, to work even though that worked in QBasic for DOS. As soon as I type them they are converted to Greek letters.

**Comment by boxgaming** [comment]
Thank you for reporting.  We’ll look at trying to address this in a future release.  In the meantime, if you are wanting to print UTF-8 characters to the screen it will display correctly if you use a font other than the default DOS font:

```basic
Dim fnt As Long
fnt = LoadFont("Arial", 18)
Font fnt
Print "Hej världen!"
```

**Comment by mobluse** [comment]
I could get your example to work but "ä" (`&auml;` in HTML) looks like greek capital letter sigma in the editor because that is still the default font. Courier works as a fixed width font, but it doesn't exactly fit with my graphics using Line and Locate no matter if I use 13 or 14. I believe the problem might be Locate. Are there more fixed width fonts that work?

```
Screen 12
Dim fnt As Long
fnt = LoadFont("Courier", 13)
Font fnt
```

**Comment by boxgaming** [comment]
There are a lot of options for loading fonts.  You can reference any standard web font, load a font from an online source (e.g. Google fonts), or load a font from a font file in the virtual file system.  Here’s a post on one of the QB64 forums with more details: https://qb64.boards.net/post/700.

Also, you might try changing the UI theme in settings.  The VSCode Dark theme and windows classic theme both use fonts that should display the characters correctly in the code editor.

**Comment by mobluse** [comment]
I couldn't find any documentation on how to store a font in the virtual file system. Maybe I could find a ttf font that looks like the default font, but uses UTF-8 instead, and then load that from the virtual file system.

The international characters work in the editor with e.g. VSCode Dark theme.

**Comment by boxgaming** [comment]
After researching further, I found that the character issue was introduced when we converted the original *.ttf formatted font file to *.woff2 format.  The missing characters have been updated and editor updates have been made to allow those characters to be displayed.  This fix will be included in the next release.  In the meantime, if you would like to preview the changes you can check them out in the development build using the github pages url:  https://boxgaming.github.io/qbjs.

As a test case you can try out an updated version of the example program that you shared there:  [nonie-updated.bas](https://boxgaming.github.io/qbjs/?code=J1RoaXMgcHJvZ3JhbSBpcyB0byBjcmVhdGUgYSB2ZXJuaWVyIGNhbGlwZXIgcmVhZGluZyBwcmFjdGljZSBnYW1lLgonaHR0cHM6Ly93d3cua3Vyc25hdmV0LnNlL8QNZXIvZnkxMjAxxwd3L3NfMV8xXzA3M19za2p1dG1hdHRfYy5odG0KSW1wb3J0IERvbSBGcm9tICJsaWIvd2ViL2RvbS5iYXMiCgpESU0gU0hBUkVEIHN2YXJUYWwsxQllbmhldCQsIMgIZmFrdG9yClNDUkVFTiAxMgpDQUxMIGluZm8KQ09OU1QgUEkgPSAzLjE0MTU5MwpkaW0gcmlnaHRyb3cgYXMgaW50ZWdlcsUYd3JvbmfMFQpETwpDTFMKUkFORE9NSVpFIFRJTUVSxmNDaG9vc2VNZXRlcgpMSU5FICgxMCwgMTAwKS0oNjPHCwrFWW1heCwgTDEsIEwyLCBzcGFjMiwgdG9sZXJhbmNlLOUBf3R5cGUsIHgxLCB5xAQyCm1heCA9IDE1J2flAaZzdCB2YWx1ZSBvZukBn2luIGNtCkwxID0gNDAnIGxvbmdlciBzY2FsZSBsaW5lCkwyxEMwJ3Nob3J0zhvlAIYgPSAyMCfEC8RPb2YgbWFpbsYmCukAoCA9IC4wMDgnY20gZXJyb3IgYWxsb3dlZAroALcgPSBJTlQoUk5EICogMS4zKSc9MCBvciAx5gD9aTLsAVNGT1LEEj0gMCBUTyAxMAogICAgeOQAwzIwMCArxBwq5gEjICogKC45ICvpAR8pJ21ha2XlAOfEX+UA1cUdxUR5xEQxMMZRSUbIZchvMTAgVEhFTsUexS/EFXkxICsgTDHFFUVMU0XINzXYLChMxTIyKSAvIDLJNyDUVscfTkQgSUbFC0NPTE9SIDE1LCDmAKTmAjTmAfIpLcYJMiksIDknNT1wdXJwbGUsIDcgZ3JleSwyIGLlArTEDmVuLDM9bMUOYmx1ZSw0PWQucmVkLDY9YnJ3OD1ncnksOT1icnQuxB4KTkVYVCBpMgpMT0NBVEUg5AKiMjYKSUbpAVXmAYpIRU7mAMVQUklOVCAiMOgA0csBMcYVMS8xMG1tIgrqAPnbNNkBMTAgMC4xxUjmASLlAitkZWws6ALeLCByY20sIHJtxgXkALrkAp7EFHht5AEsxgUyLCBtbWNtCmRlbCA9IFJORArHQCA9xE8gKuQDfQpyY23nAqbHHikKcm3PE+QCvTDFGDEw0xowKQrmAIDnAqstIMkfLeQAqynrArgxMO8C6DPnAuht5AKlx0fsAuzFHnnFHv0Cw8gLMvACzm3EFsRBLfYC0MlEygsy8ALmy0T2AujWb/EC6ecC2ecBpeQC28YL5ALdMTXwApM1LCAyN/UBTjI1CuQB2+cB10lG5QHpID4gLuYAuewCfXJjbTsgIiBjbSIg8QKZ5AH85AGTxCHkAnFlbmQgaWbLfzUy1n/lAyjddCArIDHRcCjHGcREMTDSbzf5AO7Rb9FuMthuMtVuMuQD7+YGrk51bURlYygpCsU3xGtBQlMo5wda5ACG5wMYIDzqBpogQU7mB33mB2wgPSAi5AFiT1LNPC/kApnfQcpBbW0i6gDb5QTVUm93ID3KC+QBU+oA3yJCUkEhICBEdSBoYXIgZ2rkBfMiO8ku5AHUcsKEdHQgaSByYWQh5gSXzEBGwpRyc8KUayBzZSB2YWQgZHXHRmbClCBmZWwhIvAAiuYDf+YIDD3kAeDmAtrHUCAgUsVwdsKEcmTkBT4iO8cbTEVGVCQoU1RSJCjpBHkpLCA1KTsgJyBVU0lORyAiIyMuIyMiOyDJTuQHVWxsZXIg3kjmBJzRTSMuI8k3Im1t5QOcRXhha+gAnzrkASzLUOgBpCwgOSnnAzDkAgrpAOzqAZrnBq0y6Qat5wZAMizvAZ/ClHJqYSB0aXR0YSBwwoYgaHV2dWRza2FsYW4gb3ZhbmbClHIiO8tUOchTx0IgMOQIkWplbsU/bm9uaWVu5wGV8ACF7AHTdCBodXIgbcKGbmdhIGhlbGEgY20h5QIb5QQLLuUA3Ms1U2VkYW7RNG1pbGxpbeQJw+QCEHJiacZDbW3mA5jpBDgg5AlZxl0gxF4iOyBNSUToAVbNL+QE5eQBXe8AgCAgIOYAg+oBK/0BF+UBCecBSM0p8QEP5AC5IGF0dOUC4mlsa2Vu5gFRIHNv5Ap6dGPkAzrFEuYBkjvwAJhpIGRl5ACUZmFsbMgp5gEG5gPZbeoBCOYCrGthbnPkCYnEK9Mp8APDxAsgIEzChGfHaW51bW1lcuUBBHNsdXRl5QONcuUEKSBpIGNtLuYAoklG1VvqCc3qCXjHalNrdWxsZcRmIHZhcmHnAQgxMOQB9HN0ZeQD/WF2cnVuZGEgdXBwwoZ0IHRpbGwgbsKEcm1hxCFtbecCe+cFSOoBdzfkAXbnBAsn5AfdY2/pCsU1CifKKWNvxCrFEeYAlmNv5AE6IjsKJ+QG+groA20w6QWlOSwgMTQKxj4xNMQ+SU5QVVQgIlJlZG/lAdpmb3J0c8KEdGE/IGovbiAoaikiLCByJMg05ApTJyAxMj1icmlja+QKDSAx5AnheSBkcmsgZ3J5PwpMT09QIFdISUxFIHLlBYhq5QXIxgxKywxZywx5ywwiClNZU1RFTeQGYgoKU1VC5w0AReQF2uUBKFNFTEVDVCBDQVNF5Qk95AjdUk5EKekBpcUa5gK7yAHtBhjOKeYCQ9Ip6ANY5AGMxnPlAZdTVULsAJfmDZcnZGVubmEgc3VicnV05Azaa+QC43NsdW1wYSBm5A8JxBJhIG3ChHRvbXJhYWTkAgMKJ3bChGxqxnTsDezmAOzHHmZhY+QOeMgfc2VG5g6J8ACNyRn0ASQ0/wEkxAHmDuHlBov7ASHJJuYJj8kmyg/OTMpkxScz3Sf4AWzmD2ZMU+cCieQExCwgN+oHK07kBMUuQkFTxEVUcsKEbmluZ3PoEIXmBLl2bMKEc8QYIGF25gX/5QGUIsZGyUxTdmVuc2sgVmVyc2lvbiAxOTk3IEphbuQKPCAgVFVQTyBtZWRpYSAgSmVmZiBGb3Jzc2VsbMdGICIgSMKEcm7ClHNhbmQgL1N3ZWRlbiAgVGVsOiArNDYoMCk3MC0zNSA4MCAzMDYobW9iaWwpxxo2MTEtMjIxNDQoaG9tZSkgIucA7uUEEu4ApSAg5gacYSBz5A7udCDChOoFm2FkxQ/kDXVu5gZ+5QJh5wUJ6wc4LuoAtlDChiBlxEDlEUDChnR0IChWRVJOSUVSIENBTElQRVIpIHPpBPZ0YeYE+OsG1yhtbSnLUUTlAxHkAXNlcuUJDGVub23lAKPEFGEgT1ZBTkbCmVIg5AEg5AGh5ADNKD1ub2xsKSBzdHJlY2tldMpP6Qez5waqKGtvcuQA/sUOLO0GWiwgbmVkcscp7ACT6AcZ6QbIZsKGIHRpb25kZWxzIOQN/yjkDnQgbW0pLMo25QdHciB2afEAhG9jaCBhbnbChG5k5ADk5AYM5QkNdOQHEcpBbW905AZx6gcc5AcvYsKEc3Qgc2FtbWFu5AC1ZXLkAlfpAPnwAarKTugI6y4gKE9tIHR25AbbxTXrB4JrYSBicmEsIHBywpR2YSAxLzLlD006ICdYWCxYNW1tJy4pylIg5Qr2xGhsIjvrA2HJIeUAvuYGwcRHbW0gKOUPUykgc+QApsKEcjogIDM0LjVt6ApkMy40NWNt8wLB5AIZdeQH1+gBOWEgIC7kBXbFNyAsICDlAR5kZWNpbWFsdOQBDG4uyDNvY2tzwobqALloYSBtZWxsYW5zbGFnxwvkAIksNSDkCqPESHXkCKlsLiBDTeUBrk1N5QMa5BIjIGtvcnJla3TpAx3KWG3kA7JjxEJjaCAgbW0uIE9CU+UCMewKdOQAkuUBGWnkAkBhZCAxIDIgMyAu5goRyEzmANkxMOQPSTMwxB5tbSAhIETkCItpbm5zIDIgb+UBsWzChG5n5AJI7QSd5wD07QLFIE9t6BVfbWV05gDNc8KE5Ao5QlJB5QD3aeQDynZhcizmA1rkDJZo5QoOxBrKTG1pc3N0YWcuxFLERywgc2tyaXYgbucCz2XkC+Jh5ALWxA7kAQ7JWsUOc2tpY2voBUHlAytsbOUFFy7lAOHpASLkBFnnALJmZWznARRuwoZnb3TkAaPEe+cEA8KGciDChOwEmcKO5QQ/b3PChGtlciwgdHJ5Y2sgYsRXRU5URVLlAoR2aXNhcyDnBElrbGFyaW5nLiBMeWNr5gK48QFA5wIZIEzCjk1OQewBT251LCBUxWQ8TD7lAOY8xWg+LiBBbm5hcnPoAtXmCUHJV+YAmMcs6wRj5wlR5AJo5wD3hOUA0eUA9XblCkDlDizmCDJnwpRyYSEiO+YV6OYJnkwk5A07VeQHPiQoTCQp5AicTOsO3OoJL+QKSe0HN+kPjMQoJ+YFI2RlbOgHGGjChG3kBNxpbuQBPHN0b3JsZWvmBIhpZmZlcucOe+QBBeYIo8VHxA/kAvJy5Arb5QIoaeUB4uQI3GV0YWzHMMYzc3N05AeOZ8U76wDVIkHlB4/oDYcgKEzmATRMwoRtbmEpICLmF4Qk6Aug5wD/xRTsAQLqC5rnAQfEC+sLVCdjaGFuZ2VzIGNvbW1h5BiM5wRccG9pbnQg5AelcmVtb3Zlc+UTIGVzxTbkFCNOeURlYyTkFlJzdOQCUMkZ5AVl0BjkC6hudeQSQjHkC6lMRU7nAKvpAJ/GNj3mDc7FHCzEMuQNxMkkSUbJJyIs8ADgzEUi5wy57wDt0EQg3UTZQ+cA+j3ICSvGNMUg5RNUbnVt6AGYySwi0GzKGlbkBGdlX+QC8vIM9WZvcukBOnRvIGxlbijGOvIBO21pZCTHHfIBPGlmySgiLiLkF7coNDggPD0gQVNDKMUaKeUB7ssPPD0gNTcpIHRoZeYX2cQBZWxzZckNxRF4aeUDxvQBLW5leOQEMeYBDegSpj3kGT8obGVmdO0ArS0xKeYAre4SpuYBEERPyV/zAkrwARDEAeQTV1NDKOUCTstm5AET5BMZNTjlExLaIz4gNDfuFZbwAW7rAiDEAecV5tRO5wFz307RTtks1GTOPuoWA8gB7QFWyw3sArzEAe0CwMQBSUbOPOcCyM1d5wXlRW7sDEViZWjClHZlcuQE60VOSEVU6gX15QiZxCXnAu/IAewFAFbmEUHFeD/kBO/mAPHJMOwW5+QPU1VOVElM7ACrPD7oAVXnAJ9E6QecPeQT5egCkeYQFMs6O+kGJA==)

Additionally, in order to support backwards compatibility with the [Code Page 437](https://en.wikipedia.org/wiki/Code_page_437) character codes used in QBasic programs and QB64, you will notice that these characters have a different character code than their modern unicode equivalents.

Here is a handy little program to print a character code reference in QBJS:
```basic
Screen NewImage(520, 520, 32)

Const COLS = 16
Dim As Integer i, r, c
For r = 0 To 15
    Color 15
    For i = 1 To COLS
        c = i + COLS*r
        If c = 10 Or c = 13 Or c > 255 Then
            Print "    ";
        Else
            Print "  "; Chr$(c); " ";
        End If
    Next i
    Print
    
    Color 8
    For i = 1 To COLS
        Print Right$("  " + Str$(i + COLS*r), 3); " ";
    Next i
    If i < 15 Then Print
Next r
```
(If you run this in the current production version you will see the missing character codes.)

Also, here is an example of how you can use the [Console](https://github.com/boxgaming/qbjs/wiki/Supported-Keywords#console) library to print out a character by its character code to the QBJS console so that you can copy and paste it into your source without having to use the Chr$ method:
```basic
Import Console From "lib/web/console.bas"

Console.Echo Chr$(142) + Chr$(143) + Chr$(132) + Chr$(134) + Chr$(153) + Chr$(148)
```

**Comment by boxgaming** [comment]
Update: The fix described above works in Chromium-based browsers, however Firefox and Safari do not display the characters in that range.  This is being researched further.

**Comment by mobluse** [comment]
Great! But one can still not type in ÅÄÖååö in QBJS. I use English UK/GB keyboard layout in Linux and then you type Alt+Shift+[ followed by Shift+A for Å, Alt+[ followed by Shift+A for Ä, Alt+[ followed by Shift+O for Ö, Alt+Shift+[ followed by A for å, Alt+[ followed by A for ä, Alt+[ followed by O for ö.

**Comment by boxgaming** [comment]
I decided to take a different approach altogether.  I’ve updated the default font to use standard UTF character mapping.  For backwards compatibility I also added a 437-to-UTF character crosswalk so that Asc and Chr$ methods will still return expected values.  There is also a utility now it the settings dialog that can convert the source from 437 code page to UTF in case code was loaded or pasted from an original QBasic or QB64 source.  I think this should allow for more consistent behavior now regardless of the font used.

The changes can be previewed in the development build: https://boxgaming.github.io/qbjs.

**Comment by mobluse** [comment]
Yes, I think this way is better. I tried the utility by loading a file in CP437, but when I clicked the button 437 to UTF nothing happened. I converted the file using `iconv -f CP437 -t UTF8 Nonieega.bas -o Nonieega-utf8.bas` and then loaded.

**Comment by mobluse** [comment]
There are some errors in the last 32 characters of the 255 first. This was discovered using this program:
```
cls
for i=0 to 255
if i<>0 and i<>10 then print chr$(i); else print " ";
if i mod 32=31 then print
next
```
See:
![2024-09-26-025208_814x332_scrot](https://github.com/user-attachments/assets/62633b4b-7fc7-4173-82c3-7e8566cd5427)
and compare to https://en.wikipedia.org/wiki/Code_page_437:
![250px-Codepage-437](https://github.com/user-attachments/assets/92e70764-0c77-40b9-ab75-da525dd19340)
The 3rd character in the last line should be Gamma, but is pi. The 4th character should be pi, but is something else. The 7th from the end should be a bullet, but is a centered dot. The 6th character from the end is correctly a centered dot.

**Comment by boxgaming** [comment]
Good catch.  I have fixed the character mapping issues.  The changes can be previewed in the development build: https://boxgaming.github.io/qbjs.  Let me know if you see any other issues.
